Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/BayesQuasi2.py

```diff
@@ -0,0 +1,306 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2023 ISIS Rutherford Appleton Laboratory UKRI,
+#   NScD Oak Ridge National Laboratory, European Spallation Source,
+#   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+# SPDX - License - Identifier: GPL - 3.0 +
+# pylint: disable=invalid-name,too-many-instance-attributes,too-many-branches,no-init,redefined-builtin
+
+from mantid.api import AlgorithmFactory, MatrixWorkspaceProperty, PropertyMode, Progress
+from mantid.kernel import StringListValidator, Direction
+from mantid import logger
+from IndirectCommon import GetThetaQ
+
+from typing import Dict
+from numpy import ndarray
+import numpy as np
+
+from quickBayesHelper import QuickBayesTemplate
+
+try:
+    from quickBayes.utils.general import get_background_function
+except (Exception, Warning):
+    import subprocess
+
+    print(
+        subprocess.Popen(
+            "python -m pip install -U quickBayes==1.0.0b14",
```

Comment:
These `pip install` calls merit some further discussion. which we started in the DEVS standup today.

Summarising what was mentioned there:

- This branch requires some new dependencies that are only available via `pip install` (i.e. no conda-forge package exists)
- Additional/optional dependencies have been encountered before, for machine-learning features and for force-constants import with Euphonic. We should agree a consistent approach rather than reinvent the wheel each time.
- A downside of optional dependencies is that they are not available to CI during testing. One workaround for this was used for the Euphonic force constants: a sytemtest installed Euphonic to a temporary directory, added it to the module path, and cleaned up after testing. However, this approach was a bit fragile and complex.
- At runtime, the code simply raised an error if the library was unavailable and instructed users to install it.
  - The library can be pre-installed on IDAaaS to reduce the number of users annoyed by this
- There are some things that could go wrong with a system call to `python -m pip`.
  - It is closely linked to the user environment (via `shell=True`) and this is quite variable; we can't test all possibilities. On some systems `python` may even be linked to Python 2.
  - The `-U` "upgrade" could lead to suprises.
  - No `--user` will be a problem if using a system Python where the user cannot write to default site-packages

Reply:
The best long-term option would be to create and maintain a conda-forge package for each of these dependencies, but this is a bit of extra maintenance effort/commitment for the Mantid team. (It would be polite to also reach out to the main package authors.)

Reply:
The plan is to add quickBayes to Conda in the coming months and this is designed to be a short term fix. I am trying to avoid anything that removes testing or creates more overhead for adding the library. I am going to create a more permanent solution for the next release